File: src/main/java/com/vc/easy/L437.scala
package com.vc.easy


class TreeNode437(var _value: Int) {
  var value: Int = _value
  var left: TreeNode437 = null
  var right: TreeNode437 = null
}

object L437 {
  def pathSum(root: TreeNode437, sum: Int): Int = {
    import scala.collection.mutable
    val map = new mutable.HashMap[Int, Int]()
    map.put(0, 1)
    var res = 0
    val target = sum
    def solve(root: TreeNode437, sumAcc:Int): Unit = {
      if(root != null) {
        val sumNew = sumAcc + root.value
        res += map.getOrElse(sumNew - target, 0)
        map.put(sumNew, map.getOrElse(sumNew, 0) + 1)
        solve(root.left, sumNew)
        solve(root.right, sumNew)
        map.put(sumNew, map(sumNew) - 1)
      }
    }
    solve(root, 0)
    res
  }
}

File: src/main/java/com/vc/easy/L437.java
package com.vc.easy;

import com.vc.hard.TreeNode;

import java.util.HashMap;

class L437 {
    private int res = 0, target;
    public int pathSum(TreeNode root, int sum) {
        HashMap<Integer, Integer> map = new HashMap<>();
        this.target = sum;
        map.put(0, 1);
        helper(root, 0, map);
        return res;
    }

    private void helper(TreeNode root, int sum, HashMap<Integer, Integer> map) {
        if(root == null) return;

        int total = sum + root.val;
        if(map.containsKey(total - target)) res += map.get(total - target);

        map.put(total, map.getOrDefault(total, 0) + 1);
        helper(root.left, sum + root.val, map);
        helper(root.right, sum + root.val, map);
        map.put(total, map.get(total) - 1);
    }
}
